pkg/kvstore/store/watchstore.go

```diff
@@ -0,0 +1,236 @@
+	rws := &restartableWatchStore{
+		source:     clusterName,
+		keyCreator: keyCreator,
+		observer:   observer,
+
+		state: make(map[string]*rwsEntry),
+
+		log:           log,
+		entriesMetric: metrics.NoOpGauge,
+	}
+
+	for _, opt := range opts {
+		opt(rws)
+	}
+
+	rws.log = rws.log.WithField(logfields.ClusterName, rws.source)
+	return rws
+}
+
+// Watch starts watching the specified kvstore prefix, blocking until the context is closed.
+// It might be executed multiple times, granted that the previous execution already terminated.
+func (rws *restartableWatchStore) Watch(ctx context.Context, backend WatchStoreBackend, prefix string) {
+	// Append a trailing "/" to the prefix, to make sure that we watch only
+	// sub-elements belonging to that prefix, and not to sibling prefixes
+	// (for instance in case the last part of the prefix is the cluster name,
+	// and one is the substring of another).
+	if !strings.HasSuffix(prefix, "/") {
+		prefix = prefix + "/"
+	}
+
+	rws.log = rws.log.WithField(logfields.Prefix, prefix)
+	syncedMetric := metrics.KVStoreInitialSyncCompleted.WithLabelValues(
+		kvstore.GetScopeFromKey(prefix), rws.source, "read")
+
+	rws.log.Info("Starting restartable watch store")
+	syncedMetric.Set(metrics.BoolToFloat64(false))
+
+	if rws.watching.Swap(true) {
+		rws.log.Panic("Cannot start the watch store while still running")
+	}
+
+	defer func() {
+		rws.log.Info("Stopped restartable watch store")
+		syncedMetric.Set(metrics.BoolToFloat64(false))
+		rws.watching.Store(false)
+	}()
+
+	// Mark all known keys as stale.
+	for _, entry := range rws.state {
+		entry.stale = true
+	}
+
+	// The events channel is closed when the context is closed.
+	watcher := backend.ListAndWatch(ctx, prefix, prefix, 0)
+	for event := range watcher.Events {
+		if event.Typ == kvstore.EventTypeListDone {
+			rws.log.Debug("Initial synchronization completed")
+			rws.drainKeys(true)
+			syncedMetric.Set(metrics.BoolToFloat64(true))
+
+			if !rws.synced {
+				rws.synced = true
+				for _, callback := range rws.onSyncCallbacks {
+					callback(ctx)
+				}
+			}
+
+			continue
+		}
+
+		key := strings.TrimPrefix(event.Key, prefix)
+		rws.log.WithFields(logrus.Fields{
+			logfields.Key:   key,
+			logfields.Event: event.Typ,
+		}).Debug("Received event from kvstore")
+
+		switch event.Typ {
+		case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+			rws.handleUpsert(key, event.Value)
+		case kvstore.EventTypeDelete:
+			rws.handleDelete(key)
+		}
+	}
+}
+
+// NumEntries returns the number of entries synchronized from the store.
+func (rws *restartableWatchStore) NumEntries() uint64 {
+	return rws.numEntries.Load()
+}
+
+// Drain emits a deletion event for each known key. It shall be called only
+// when no watch operation is in progress.
+func (rws *restartableWatchStore) Drain() {
+	if rws.watching.Swap(true) {
+		rws.log.Panic("Cannot drain the watch store while still running")
```

Comment:
The idea of being strict here was to easily catch possible misuses, since draining it while actively watching a kvstore would lead to inconsistent behaviors (both due to the meaning of the operation itself and being the local state not protected by a mutex for the sake of simplicity).